### PR TITLE
Use index when `a=<literal>` gets an implicit cast

### PR DESCRIPTION
**Pull request by ggevay**
If we have `<expr> = <literal>` where an implicit cast is inserted on `<expr>`, then an index on `<expr>` will not be used. For example, let's say we have an index on a SMALLINT column `a`, and we do `WHERE a = 5`. The overload resolution for the `=` will insert a cast:
`smallint_to_integer(a) = 5`, and then the literal constraint detection won't find `a`'s index.

We have the same issue if `a` is a VARCHAR column, and we do `WHERE a = 'xxx'`. (The situation here is harder to work around than the SMALLINT example above, because even `WHERE a = 'xxx'::VARCHAR` won't work, because we don't have an `=` implementation for VARCHAR. The only workaround currently is to create the index on `a::VARCHAR`.)

This PR solves some common cases of this problem by moving the cast from the indexed column to the literal: We do
```
<lit> = f(<expr>), where f is invertible
 -->
<f^-1(lit)> = <expr>
```
There are two shortcomings of this PR:
1. We only do the transform if `f^-1(lit)` doesn't error out. This means that if we have `WHERE a = 3000000000` and a SMALLINT index on `a`, then we do an index scan instead of realizing that this is an impossible filter.
2. The bigger problem is that finding a cast function's inverse in general currently seems a bit hard, so I just hardcoded some common cases for now. @sploiselle is thinking about a more general solution.

I think we should quickly merge this before tomorrow's release even if the above shortcomings are not solved until then, and then we'll have time to think about a more general solution until next week.

### Motivation

  * This PR fixes a recognized bug:
    * https://github.com/MaterializeInc/materialize/issues/15525
    * #10738
    * another user running into this: https://materializeinc.slack.com/archives/C02PPB50ZHS/p1665769455773719,
    * older discussion: https://materializeinc.slack.com/archives/CMHDK0DK8/p1645629475658999
    * (Also occurs in TPC-H Q08)

(Note that this is related also to https://github.com/MaterializeInc/materialize/issues/4171, but that cannot be fixed by casting some magic spell inside `MirScalarExpr::reduce`, since a cast on either the left-hand side or the right-hand side of a join equality would prevent index usage in one of the directions of a Delta join. To fix that, we'll need to do an inversion inside the  join planning.)

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - Use `a`'s index even when an implicit cast is needed in `a = <literal>`.

### CODE REVIEW

**Review by sploiselle — APPROVED** [review]
Makes sense to me to merge this if there's urgency. A general solution is hard, so might just propose a more structured version of this approach as a v1

File: test/sqllogictest/literal_constraints.slt
Comment (sploiselle) [inline test/sqllogictest/literal_constraints.slt:758]: Should we also add a test/support for the literal on the RHS?

Comment (ggevay) [reply test/sqllogictest/literal_constraints.slt:771]: Yes, absolutely, thx! Done.

This case is handled by the canonicalization in `MirScalarExpr::reduce` around here:
```
                        } else if matches!(*func, BinaryFunc::Eq | BinaryFunc::NotEq
                            && expr2 < expr1
                        {
                            // Canonically order elements so that deduplication works better.
                            // Also, the below `Literal([c1, c2]) = record_create(e1, e2)` matching
                            // relies on this canonical ordering.
                            mem::swap(expr1, expr2);
                        }
```

File: src/expr/src/scalar/mod.rs
Comment (aalexandrov) [inline src/expr/src/scalar/mod.rs:893]: nit: it's a bit easier to follow if you use consistent naming of meta-variables in the docs and the Rust variables in the actual code. For example here the Rust variables are `literal`, `func`, `expr`, but in the code they are `lit`, `f`, `expr`.

File: src/expr/src/scalar/mod.rs
Comment (aalexandrov) [inline src/expr/src/scalar/mod.rs:902]: nit: `inved_lit` ⇒ `inverted_lit` or `lit_inv` is probably more obvoius?

File: src/expr/src/scalar/mod.rs
Comment (aalexandrov) [inline src/expr/src/scalar/mod.rs:904]: I think you can do `take_dangerously()` instead of `clone()` here (assuming this is defined on `MirScalarExpr`).

File: src/expr/src/scalar/mod.rs
Comment (aalexandrov) [inline src/expr/src/scalar/mod.rs:921]: Same as above: no need to `clone` as the `expr` is going to be dropped at the end of this block, so you can just take ownership.

**Review by aalexandrov — APPROVED** [review]
Looks good, but let's wait for @philip-stoev's review before we merge this.

**Comment by philip-stoev** [comment]
I would like to do some experimenting around this, please give me 2 hours.

**Comment by ggevay** [comment]
Ok, thanks @philip-stoev! I'm addressing @aalexandrov's comments now, will push an update in a few minutes, maybe you could wait with the testing until then.

**Comment by ggevay** [comment]
Pushed the update, @philip-stoev

File: src/expr/src/scalar/mod.rs
Comment (ggevay) [reply src/expr/src/scalar/mod.rs]: Done

File: src/expr/src/scalar/mod.rs
Comment (ggevay) [reply src/expr/src/scalar/mod.rs]: Done. (It's just called `take` for some reason)

File: src/expr/src/scalar/mod.rs
Comment (ggevay) [reply src/expr/src/scalar/mod.rs]: changed

File: src/expr/src/scalar/mod.rs
Comment (ggevay) [reply src/expr/src/scalar/mod.rs]: fixed

**Comment by ggevay** [comment]
It seems that the macro stuff that @benesch mentioned on Slack is not trivial for me to do, I would need a few hours (I never wrote or modified Rust macros before). So I'd say let's merge this after @philip-stoev's testing to make the release cut, and then I'll make it nice with macros in a follow-up PR.

**Comment by philip-stoev** [comment]
item No 1 . This particular case does not use an index (the reverse case does).

```
drop table if exists dec_table;
create table dec_table (f1 integer);
create index i1 on dec_table(f1);
explain select * from dec_table where f1 = 2::decimal(5,2);

               Optimized Plan                
---------------------------------------------
 Explained Query (fast path):               +
   Filter (2 = integer_to_numeric(#0))      +
     ReadExistingIndex materialize.public.i1+
                                            +
 Used Indexes:                              +
   - materialize.public.i1                  +
```

The raw plan has a weird cast on both sides:

```
materialize=> explain raw plan for select * from dec_table where f1 = 2::decimal(5,2);
                        Raw Plan                         
---------------------------------------------------------
 Filter (integer_to_numeric(#0) = integer_to_numeric(2))+
   Get materialize.public.dec_table                     +
```

**Comment by philip-stoev** [comment]
Item No 2 . If a person has already defined an index using a cast in the indexed expression, that index would not be used:

```
drop table if exists t1;
create table t1 (f1 varchar);
create index i1 on t1(f1::text);
explain select * from t1 where f1 = 'a'::text;
explain select * from t1 where f1::text = 'a';
explain select * from t1 where f1::text = 'a'::text;
```

Not sure if this is important or even fixable ... just noting this in case a customer has already applied a self-created workaround to the problem at hand.

All of the 3 queries above would use an index lookup in `main`.

**Comment by philip-stoev** [comment]
Item No 3. This will not use an index (the reverse case does):

```
drop table if exists t1;
create table t1 (f1 timestamp);
create index i1 on t1(f1);
explain select * from t1 where f1 = CAST('2011-11-11 11:11:11' AS timestamp with time zone);

                                 Optimized Plan                                 
--------------------------------------------------------------------------------
 Explained Query (fast path):                                                  +
   Filter (2011-11-11 11:11:11 UTC = timestamp_to_timestamp_with_time_zone(#0))+
     ReadExistingIndex materialize.public.i1                                   +
                                                                               +
 Used Indexes:                                                                 +
   - materialize.public.i1                                                     +

Same as before, the raw plan looks like this:
```
 Filter (timestamp_to_timestamp_with_time_zone(#0) = text_to_timestamp_with_time_zone("2011-11-11 11:11:11"))+
   Get materialize.public.t1 
```

I am not sure if the cast from timestamp to timestamp with tz and back is reversible or not. @mjibson would probably know best.

**Comment by philip-stoev** [comment]
Item No 4 . Unsigned types do not seem to be covered :+1: 
```
drop table if exists t1;
create table t1 (f1 uint2);
create index i1 on t1(f1);
explain select * from t1 where f1 = 1::integer;
               Optimized Plan                
---------------------------------------------
 Explained Query (fast path):               +
   Filter (1 = uint2_to_integer(#0))        +
     ReadExistingIndex materialize.public.i1+
                                            +
 Used Indexes:                              +
   - materialize.public.i1                  +
```

even when it would be safe in terms of overflow

**Comment by philip-stoev** [comment]
Item No 5. The index lookup does not kick in in this case:

```
materialize=> create table t1 (f1 date);
CREATE TABLE
materialize=> create index i1 on t1(f1);
CREATE INDEX
materialize=> explain select * from t1 where f1 = '2011-11-11'::timestamp;
                     Optimized Plan                     
--------------------------------------------------------
 Explained Query (fast path):                          +
   Filter (2011-11-11 00:00:00 = date_to_timestamp(#0))+
     ReadExistingIndex materialize.public.i1           +
                                                       +
 Used Indexes:                                         +
   - materialize.public.i1                             +

```

The RAW plan has two distinct casts:
```
materialize=> explain raw plan for select * from t1 where f1 = '2011-11-11'::timestamp;
                             Raw Plan                             
------------------------------------------------------------------
 Filter (date_to_timestamp(#0) = text_to_timestamp("2011-11-11"))+
   Get materialize.public.t1                                     +
 ```

(I did not explore the full matrix of timestamp/date/time types)

**Review by philip-stoev — APPROVED** [review]
Thanks. Seems to work as advertised apart from the items noted. I hope that most can be knocked off by extending the `match` , but note that there are some where a cast appears on both sides of the equality.

**Comment by ggevay** [comment]
@philip-stoev 
First, regarding the raw plans: Those are expected to have the casts on both sides. Constant folding removes the casts from the literals at some point.

re item No 1.:
To support this, we would need to insert a rounding cast, which in general might be surprising. E.g., what if the user writes `WHERE a = 3.14` for an integer column `a`? Does this user want us to find `a = 3`? I think not. (What we _could_ do, is check if the rounding is a no-op for the particular literal at hand, i.e., check whether doing the cast back-and-forth gives back the original literal. But I think it's not worth the trouble, because if the literal is already a round number, then it probably won't come out as a decimal or float (unless there is an explicit cast, but I don't see why would someone do that).)

re Item No 2.:
Hm, yes, this is a bit unfortunate :( I'm thinking now about how hard would it be to fix this.

re Item No 3.:
I'm not sure about this timestamp stuff, will get back to this later.

re Item No 4.:
I added support for unsigned types now.

re Item No 5.:
Time stuff -- will get back to this later.

re Item No 6.:
Is there any reason why would the literal have a char type? An explicit cast can make it a char type, but why would someone do that? Or do we have some string function that produces a char type when working with varchars, and then maybe constant folding would produce a char literal?
Maybe the reverse case is more important, i.e., when the column type is a char type, and somebody tries to do a lookup with text.

**Comment by sploiselle** [comment]
I have the generalized version of this prepped [here](https://github.com/sploiselle/materialize/tree/right-inverse); will wait for this to merge before putting up the PR.

**Comment by philip-stoev** [comment]
> To support this, we would need to insert a rounding cast, which in general might be surprising. E.g., what if the user writes WHERE a = 3.14 for an integer column a? Does this user want us to find a = 3? I think not. 

Yes, `WHERE a = 3.14` should **not** match a row with `3`. If this can not be assured, then lets' drop it. Postgres is also unable to use an index in this case, so I guess it is unsafe in general.

> Is there any reason why would the literal have a char type? An explicit cast can make it a char type, but why would someone do that? Or do we have some string function that produces a char type when working with varchars, and then maybe constant folding would produce a char literal?

You are right, this is a very remote possibility -- I do not think we have any functions that produce `char`. Even the `chr` function returns `text`.

The only realistic scenario I can think of is if one is using a prepared statement with `?` placeholders and then the database driver / ORM framework explicitly provides a `char` type when binding the actual value to be used when running the query.

> Maybe the reverse case is more important, i.e., when the column type is a char type, and somebody tries to do a lookup with text.

I do not recall if I checked the inverse case...

**Comment by ggevay** [comment]
I converted this to a draft now, because I need to make substantial changes to address Item No 2., which is to make it work also in the case when the user already applied the workaround of creating the index with the extra cast in it. I don't see how to fix this in `MirScalarExpr::Reduce` in a way that both normal indexes and workarounded indexes work, so I'll move this cast elimination code from `reduce` into the literal constraints detection (in `CanonicalizeMfp`). That will also enable properly handling the case when the inverse cast errors out.

**Comment by ggevay** [comment]
Now I ran into some additional trouble with `uint4` and `uint8`: When comparing an `integer` literal to these types, both sides of the equality are being implicitly cast to `double`:
```
materialize=>
create table foo(a uint4);
create default index on foo;
explain select * from foo where a = 5;
CREATE TABLE
CREATE INDEX
                      Optimized Plan                      
----------------------------------------------------------
 Explained Query (fast path):                            +
   Filter (5 = uint4_to_double(#0))                      +
     ReadExistingIndex materialize.public.foo_primary_idx+
```
This is a problem, because I was planning to NOT insert the inverse cast when `preserves_uniqueness` is false, and for `double_to_uint4` `preserves_uniqueness` is false.

So now I'm thinking about an alternative way to decide whether to insert the inverse cast: Instead of relying on `preserves_uniqueness`, we try to go back and forth with the reverse cast and the original cast on the literal value, and see if we get back the original literal value. If yes, then we assume that doing the inverse cast is fine. (I.e., we check that `f(f_inv(literal)) = literal` for this particular literal value.) For example, assume an `integer` column `a`, and then we would do an index lookup for `WHERE a = 3::DOUBLE` (because converting the double value 3 to integer is 3 and converting that back to double is again 3), but we wouldn't do an index lookup for `WHERE a = 3.14::DOUBLE` (because converting the double value 3.14 to integer is 3, and converting that back to double is 3, which does not equal 3.14). I think this is correct behavior. @sploiselle, @philip-stoev, do you think this makes sense?

**Comment by philip-stoev** [comment]
I think the `uint4_to_double` cast above is a potentially lossy cast, so it is a bug on its own that may need to be moved out of the way before you proceed. we should not be doing anything even remotely touching doubles if the entire context is intergers-only. 

> because converting the double value 3 to integer is 3 and converting that back to double is again 3

That is not really guaranteed I think, converting 3 back to double would return 3.00000000000000000000000000001 or whatever and your test will fail ?

**Comment by sploiselle** [comment]
@ggevay Let me check in on why this is being moved to a `double`; probably something buggy in function selection for `Eq`

**Comment by ggevay** [comment]
Maybe the implicit double cast is because Postgres also does it and we wanted to match its behavior? I'm not sure. Btw.:
```
materialize=> select 123456789::integer = 123456789::integer::double::integer;
 ?column? 
----------
 t
(1 row)

materialize=> select 123456789::integer::double = 123456789::integer::double::integer::double;
 ?column? 
----------
 t
(1 row)
```

**Comment by sploiselle** [comment]
Ah, the problem is that the type promotions between `int` and `uint` are not fleshed out/not ideal. The problem here is that the "best common type" between i32 and u32 is float64 rather than i64. Digging into this a bit further.

**Comment by sploiselle** [comment]
Ah, ok--so the issue is that we will never choose to cast `i32` and `u32` to anything other than `float64` because they are not implicitly castable to each other, but are both castable to their type category's preferred type (`float64`). In essence, this is a problem for us and not PG because we've introduced a kind of tree into our numeral types, where the signed and unsigned integers of the same size do not have parity with one another. In PG, there's a strictly linear relationship with the domain of numbers that can be presented.

However, w/r/t index access, the only meaningful way to fix this is to cast the literal to be of the same type as the column.

**Comment by ggevay** [comment]
> However, w/r/t index access, the only meaningful way to fix this is to cast the literal to be of the same type as the column.

Hm, I'm not sure how to do that in MIR. The literal was already cast to double, and constant folding already actually performed the conversion before we reach the MIR optimizations. So anything that I can do in MIR has to work with the double value, which might be imprecise according to Philip. I'm out of ideas at the moment, will think more tomorrow.

**Comment by ggevay** [comment]
It seems that 32-bit integer to 64-bit floating point is probably lossless (this is C++, I hope it's the same for Rust):
https://stackoverflow.com/a/43656339/5460002
So maybe our tree could be revised a bit to allow implicit conversion from INTEGER to REAL?
(Not true from 64-bit integer to 64-bit floating point.)

**Comment by sploiselle** [comment]
@ggevay Sorry for the ambiguity--I meant that you would have to introduce the cast into the expression in the query when entered--I don't think there's a way for the optimizer to overcome this.

**Comment by ggevay** [comment]
I completely rewrote the PR. It's ready for review, @aalexandrov. I'm thinking that @philip-stoev could test these changes not before merging this PR, but when @sploiselle later rebases his follow-up work after merging this one. The current PR works only for a few types, and Sean's work will add support for many more types.

I added many tests in `literal_constraints.slt`, and I locally ran a big portion of the nightly sqlite slt tests.

@sploiselle, I changed `CastStringToVarChar::preserves_uniqueness` as we discussed on Slack. Could you please check that line?

The main change of the rewrite is that I moved the reverse casting from `MirScalarExpr::reduce` to `CanonicalizeMfp`, i.e., into the index selection. This change was needed because there I can already see what index I'm trying to match, so I can decide whether I need to do the inverse casting or not based on how the index looks. This enables matching both the normal indexes, and also those indexes where the workaround of adding the cast at index creation was applied. (Philip's Item No 2)

The rewrite also enabled solving the 1. shortcoming mentioned in the PR description: Now we detect the literal equality as being impossible if the inverse cast fails. (Doing this in `reduce` would have been very cumbersome, because we might have `is_null` somewhere around the literal equality expression, but in `CanonicalizeMfp` the context is more controlled: the result of the literal equality will be used as a boolean.)

The 2. shortcoming mentioned in the PR description will be solved by Sean's follow-up work.

No need to check `invert` in `func.rs`, because that will be superseded by Sean's work soon.

There is one more problem that this rewrite is not solving: the unsigned types:
- (`uint2` works, because the `uint2` is upcast to `int4` in a `u2 = i4` equality.)
- `uint8` seems hopeless to solve automatically. The user can insert an explicit cast. I'll document this when I'm working on https://github.com/MaterializeInc/materialize/issues/14645
- `uint4` doesn't work, but we might be able to solve this later (it's not urgent): @sploiselle, Maybe we could modify the implicit cast insertion logic in HIR to insert a cast to `int8` instead of `double`, when we have an `u4 = i4` equality?

File: src/expr/src/scalar/func/impls/string.rs
Comment (sploiselle) [inline src/expr/src/scalar/func/impls/string.rs:475]: LGTM!

**Comment by sploiselle** [comment]
@ggevay wrt uint/int types, filed #15536; will put up a fix for that today

File: src/expr/src/scalar/mod.rs
Comment (aalexandrov) [inline src/expr/src/scalar/mod.rs:352]: I thought this was moving completely to `canonicalize_mfp`. What's the reason of keeping this method?

**Comment by ggevay** [comment]
Merged this now. @sploiselle, you can rebase your stuff. We are on track to have a complete fix in the next release for https://github.com/MaterializeInc/materialize/issues/15525

File: src/expr/src/scalar/mod.rs
Comment (ggevay) [reply src/expr/src/scalar/mod.rs:352]: This is the join issue: https://github.com/MaterializeInc/materialize/issues/4171

File: src/expr/src/scalar/func.rs
Comment (aalexandrov) [inline src/expr/src/scalar/func.rs:3402]: Maybe an example that illustrates this will be good here.

File: src/expr/src/scalar/func.rs
Comment (ggevay) [reply src/expr/src/scalar/func.rs:3776]: yes

File: src/expr/src/scalar/func/impls/string.rs
Comment (aalexandrov) [inline src/expr/src/scalar/func/impls/string.rs:474]: cc @philip-stoev as far as I understand, this is related to some of the corner cases that you were thinking about last week.

File: src/expr/src/scalar/mod.rs
Comment (aalexandrov) [reply src/expr/src/scalar/mod.rs:352]: Got it.

File: src/expr/src/scalar/func.rs
Comment (ggevay) [reply src/expr/src/scalar/func.rs:3402]: Ok, I'll extend the comment. For example, casting the integer 1000000 to smallint yields the same error result as casting the integer 2000000 to smallint.

**Pull request merged**
